ground-server/src/app/page.tsx

```diff
@@ -17,31 +17,45 @@ const ResponsiveReactGridLayout = WidthProvider(Responsive);
 export default function Home() {
   const [widgets, setWidgets] = useState<Widget[]>([]);
   const [data, setData] = useState<Data[]>([]);
+  const [ws, setWs] = useState<WebSocket | null>(null); // Store the WebSocket instance
```

Comment:
One thing is you typically use `useRef` for stuff that shouldn't cause a rerender on change (like the websocket instance). useState always triggers a rerender and there is no need to do that when setting the websocket

Reply:
Oh neat! Thanks for pointing that out